htlcswitch/circuit_map.go

```diff
@@ -786,22 +788,15 @@ func (cm *circuitMap) DeleteCircuits(inKeys ...CircuitKey) error {
 	)
 
 	cm.mtx.Lock()
-	// First check that all provided keys are still known to the circuit
-	// map.
+	// Remove any references to the circuits from memory, keeping track of
+	// which circuits were removed, and which ones had been marked closed.
+	// This can be used to restore these entries later if the persistent
+	// removal fails.
 	for _, inKey := range inKeys {
-		if _, ok := cm.pending[inKey]; !ok {
-			cm.mtx.Unlock()
-			return ErrUnknownCircuit
+		circuit, ok := cm.pending[inKey]
```

Comment:
This change is primarily in place in order to address _duplicate_ HTLC forwards, correct? I guess also the case of a set of items in a forwarding package that has cancels only being _partially_ processed?

Reply:
Main thing is that this allows deletion to be idempotent. It will succeed if the item gets deleted or already has been. We need this for two reasons:
1) in case we want to delete circuits for spurious responses, as we may need to delete twice
2) the split mailbox issue caused circuits not to be deleted in the same batches as they were committed

So this enables 1 by allowing duplicate deletion, and is more forgiving to the second case by making it best effort

Reply:
👍